### PR TITLE
Enable realtime inference in AutoMM

### PR DESCRIPTION
**Pull request by zhiqiangdon**
1. Enable realtime inference without trainer. This can reduce 10x inference time in the online deployment scenario where each time only a small batch comes for inference.
2. Add unit tests.
3. Add one util to move objects to device.
4. Move `get_collate_fn ` outside of `datamodule` to make it easier to use in more places.

Below are some initial performance comparisons with the electra-base model on the SST dataset.

Default prediction time:

batch size 1, inference time 0.6611239910125732
batch size 2, inference time 0.6466214656829834
batch size 3, inference time 0.6647155284881592
batch size 4, inference time 0.6407644748687744
batch size 5, inference time 0.6691045761108398
batch size 6, inference time 0.6658930778503418
batch size 7, inference time 0.6808614730834961
batch size 8, inference time 0.6343615055084229
batch size 9, inference time 0.647843599319458

batch size 100, inference time 0.7578756809234619
batch size 101, inference time 0.7177624702453613
batch size 102, inference time 0.7434618473052979
batch size 103, inference time 0.7381682395935059
batch size 104, inference time 0.7458250522613525
batch size 105, inference time 0.7460315227508545
batch size 106, inference time 0.7509653568267822
batch size 107, inference time 0.7573814392089844
batch size 108, inference time 0.764272928237915

Realtime prediction time:

batch size 1, inference time 0.0407717227935791
batch size 2, inference time 0.04250025749206543
batch size 3, inference time 0.04233431816101074
batch size 4, inference time 0.043201446533203125
batch size 5, inference time 0.04405808448791504
batch size 6, inference time 0.04470324516296387
batch size 7, inference time 0.0460968017578125
batch size 8, inference time 0.04686594009399414
batch size 9, inference time 0.04734015464782715

batch size 100, inference time 0.13451266288757324
batch size 101, inference time 0.1346583366394043
batch size 102, inference time 0.13354206085205078
batch size 103, inference time 0.13328337669372559
batch size 104, inference time 0.13449764251708984
batch size 105, inference time 0.13887500762939453
batch size 106, inference time 0.14088964462280273
batch size 107, inference time 0.14155197143554688
batch size 108, inference time 0.14289355278015137

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by sxjscience — APPROVED** [review]
LGTM. Next time, when we are changing the logic of our LightningModule, we need to also remember to change the logic of `_realtime_predict`.

**Review by yinweisu — APPROVED** [review]
LGTM! Awesome feature!

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 1]
```diff
@@ -1524,6 +1595,8 @@ def evaluate(
             If None, we only return the score for the stored `_eval_metric_name`.
         return_pred
             Whether to return the prediction result of each row.
+        realtime
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1598]: Good point! Working on it.

**Comment by zhiqiangdon** [comment]
> LGTM. Next time, when we are changing the logic of our LightningModule, we need to also remember to change the logic of `_realtime_predict`.

Good suggestion! I will have follow-up PRs to try to reuse some logic.

**Review by bryanyzhu — APPROVED** [review]
LGTM. Just one question, how small counts for small data? If you can give a concrete example, it will help users to decide whether to turn on this flag or not.

**Comment by zhiqiangdon** [comment]
> LGTM. Just one question, how small counts for small data? If you can give a concrete example, it will help users to decide whether to turn on this flag or not.

I have posted some initial time comparisons. I would say a 

> LGTM. Just one question, how small counts for small data? If you can give a concrete example, it will help users to decide whether to turn on this flag or not.

I have posted some comparisons in the description. We can see that the realtime inference is still several times faster even for batch size 100.

**Comment by sxjscience** [comment]
Should we just remove the Trainer logic and write our own inference code then?

**Comment by zhiqiangdon** [comment]
> Should we just remove the Trainer logic and write our own inference code then?

The default prediction is still useful if we have many samples (>1000) for inference.

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 2]
```diff
@@ -1505,11 +1507,80 @@ def evaluate_coco(
...
+            for per_modality, per_modality_processors in data_processors.items():
+                for per_model_processor in per_modality_processors:
+                    if modality_features[per_modality]:
+                        per_item_features.update(
+                            per_model_processor(modality_features[per_modality], idx=i, is_training=False)
+                        )
+            processed_features.append(per_item_features)
+
+        collate_fn = get_collate_fn(df_preprocessor=df_preprocessor, data_processors=data_processors)
+        batch = collate_fn(processed_features)
```
Comment (FANGAreNotGnu) [inline multimodal/src/autogluon/multimodal/predictor.py:1543]: Here is the batch we need for ONNX export/evaluation. We can modularize code above in next pr. For now I'll copy this code into export_onnx.

**Review by FANGAreNotGnu — APPROVED** [review]
LGTM, thanks for the feature! It enables the evaluation for ONNX model (may be also helpful to our multipy integration).

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 3]
```diff
@@ -1505,11 +1507,80 @@ def evaluate_coco(
...
+            requires_label=requires_label,
+        )
+
+        lengths = []
+        modality_features = {}
+        for per_modality in data_processors:
+            per_modality_features = getattr(df_preprocessor, f"transform_{per_modality}")(data)
+            modality_features[per_modality] = per_modality_features
+            if per_modality_features:
+                lengths.append(len(per_modality_features[next(iter(per_modality_features))]))
```
Comment (suzhoum) [inline multimodal/src/autogluon/multimodal/predictor.py:1527]: Could you help me understand what `per_modality_features[next(iter(per_modality_features))]` does here?

File: multimodal/src/autogluon/multimodal/predictor.py [hunk 4]
```diff
@@ -1505,11 +1507,80 @@ def evaluate_coco(
...
+                for per_model_processor in per_modality_processors:
+                    if modality_features[per_modality]:
+                        per_item_features.update(
+                            per_model_processor(modality_features[per_modality], idx=i, is_training=False)
+                        )
+            processed_features.append(per_item_features)
+
+        collate_fn = get_collate_fn(df_preprocessor=df_preprocessor, data_processors=data_processors)
+        batch = collate_fn(processed_features)
+        precision = infer_precision(num_gpus=1, precision=self._config.env.precision, as_torch=True)
```
Comment (suzhoum) [inline multimodal/src/autogluon/multimodal/predictor.py:1544]: should `num_gpus` be hard coded?

File: multimodal/tests/unittests/test_predictor.py [hunk 5]
```diff
@@ -59,6 +59,22 @@ def verify_predictor_save_load(predictor, df, verify_embedding=True, cls=MultiMo
             assert embeddings.shape[0] == len(df)
 
 
+def verify_realtime_inference(predictor, df, verify_embedding=True):
```
Comment (suzhoum) [inline multimodal/tests/unittests/test_predictor.py:62]: Just curious, will there be a case when we don't want to verify embedding?

Comment (zhiqiangdon) [reply multimodal/tests/unittests/test_predictor.py:62]: It seems that extracting embedding may be used in all kinds of pipelines and models.

File: multimodal/src/autogluon/multimodal/predictor.py (hunk 3 above)
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/predictor.py:1527]: Each modality may have multiple columns in dataframe, which corresponds to the keys in the `per_modality_features `. Here want to compute the sample number by only using the first key in the dict.

Comment (suzhoum) [reply multimodal/src/autogluon/multimodal/predictor.py:1527]: That makes sense now. Thanks!

**Pull request merged**